ENH: Add annotations for `np.linalg`

BvB93 opened the pull request:
This PR adds annotations for the `np.linalg` sub-package.

All in all I was pleasantly suprised by the functions' signatures, as adding dtype-support was made considerably easier by the minimal amount of ndarray-scalar conversions, the lack of `dtype` parameters and the (relativelly) small range of supported dtypes.

Commit:
MAINT: Move `linalg` function definitions to `linalg.linalg`

Commit:
ENH: Add annotations for `np.linalg`

Commit:
TST: Add typing tests for `np.linalg`

bashtage commented on numpy/linalg/linalg.pyi:
```diff
@@ -0,0 +1,274 @@
+def eigvalsh(a: _ArrayLikeComplex_co, UPLO: L["L", "U", "l", "u"] = ...) -> NDArray[floating[Any]]: ...
+
+@overload
+def eig(a: _ArrayLikeInt_co) -> _2Tuple[NDArray[float64]]: ...
```
Not sure if this is correct
```
In [7]: np.linalg.eigvals(np.array([[-3,2,1],[-1,-3,1],[1,-2,-3]],dtype=int))
Out[7]: array([-2. +0.j        , -3.5+1.93649167j, -3.5-1.93649167j])
```

BvB93 replied:
You're right, apparently I was previously hitting the branch of code below with `all(w.imag == 0)`.
This is rather unfortunate, as I don't see any way of typing value-based casts...

https://github.com/numpy/numpy/blob/5d75d4862fd3409b54a29a3523e60ac1d6e374ec/numpy/linalg/linalg.py#L1059-L1063

BvB93 replied:
> This is rather unfortunate, as I don't see any way of typing value-based casts...

Actually, scratch that: this would be one of the rare situation where returning a union _is_ appropiate.

bashtage replied:
This applies to all of the eigenvalue functions I think.

BvB93 replied:
Fixed as of https://github.com/numpy/numpy/pull/19887/commits/f8958a2305c88f1c97b2cd5256d7e29faca93f13.

Commit:
MAINT: Let `eig` and `eigvals` return a union for real inputs

Whether an array of a `floating` or `complexfloating` dtype is returned depends on the values within in the supplied matrix.

Co-Authored-By: Kevin Sheppard <bashtage@users.noreply.github.com>

charris commented:
Thanks Bas.

tupui commented:
I think this change is causing some issues in our nightly CI. We have `from numpy.linalg.linalg import LinAlgError` and this is now failing.

BvB93 commented:
Ah, seems I forgot to re-export `LinAlgError`. https://github.com/numpy/numpy/pull/19909 should fix the issue, but just as a general remark: the `np.linalg.linalg` module is apparently [not considered public API](https://github.com/numpy/numpy/blob/50785ae68e95296c9afbe62c3f50e5d9c6b71c64/numpy/tests/test_public_api.py#L274), so it might be worthwhile to import the exception from `numpy.linalg` instead.

tupui commented:
Thanks! I will see to fix this on SciPy's side as well.
